### ISSUE TITLE
The influence of adding MPL for improving the performace

### ISSUE THREAD

**Issue by ChillTerry** [qherau/3DGS-Calib#3]
Hi Quentin Herau,
   It's a great work for online calibration. After reading the paper, I’m curious about the importance of adding the MLP in this work. Have you conducted an ablation study on the extrinsic calibration without the MLP? How much the MLP improves the final accuracy of extrinsic?

**Comment by qherau** [comment]
Hi,
Thank you for your kind words. Essentially, without the MLP, convergence was extremely slow. After a full day of training without achieving convergence, I had to terminate the process. The underlying intuition was that each time we updated the poses, every Gaussian needed to optimize its parameters individually. By incorporating the MLP, we enable the sharing of parameters among spatially close Gaussians, which significantly enhances efficiency.
Therefore, the decision was driven by the need for speed rather than precision.

**Comment by ChillTerry** [comment]
Thank you for your active reply!
But I still have a question about the speed issue. To my knowledge, for the gaussians in the original 3dgs paper also need to optimize  its parameters individually. Thanks to the CUDA implementation and advanced off-the-shelf rendering pipeline, it can achive fast training time. 
For our extrinsic calibration purpose, from my perspective, we just need to change the original Tcw to Tcw*Topt (where Topt is setted to the optimzed variable) to rectify the original wrong extrinsic (Tcw), then the reset of things should be the same to implementation of original 3dgs paper. It might increase optimization time due to introducing new optimization variable Topt, but it should not take a whole day?

**Comment by qherau** [comment]
What is long is not to train the Gaussians per se, it is about the same speed as the original 3DGS. The problem is that because the Gaussians are static, and the extrinsics are changing, the information that the Gaussian is trying to learn changes. For example, in the normal training, a Gaussian becomes red and that is it. In our case, it might try to be green, then blue, then red as the poses of the images change. And each individual Gaussian would have to independently change accordingly.

**Comment by ChillTerry** [comment]
Thx! 
This deep talk leads to my rethinking about the optimization scheme. As we know that the original 3dgs paper choose CUDA for acceleration. So the authors explicitly derived the gradients of a scalar loss with respect to the input Gaussian parameters. If we optimize the extrinsic by modifying Tcw (this's just my assumption that mentioned above), then we should  explicitly derived the gradient of loss w.r.t. Tcw as well. But from the 3dgs-calib paper I haven't seen something about this. Do you choose another way to achive this in code?

**Comment by ChillTerry** [comment]
Hi, Herau. It’s been a few days. Maybe you missed this issue, so I’m sending this message to kindly remind you about our discussion and look forward to your reply.

**Comment by qherau** [comment]
Hi, sorry for the late reply, I didn't see the notice. As I'm not very familiar with CUDA, and there was no code at the time to propagate the gradients back to the poses, what I did was to transform all the Gaussian positions to the camera referential, using (Tcw)^-1, which allowed me to add the pose correction to the gradients without changing the CUDA code, and let pytorch take care of that.

**Comment by ChillTerry** [comment]
Thank you for your reply! 

I think we can expand the $T_{cw}$ a little bit for a more detailed explanation. Before that, we should form a consensus that $T_{cw}$ means the transformation from world to camera. Because I see you use $T_{cw}^{-1}$ to transform all the Gaussian positions to the camera referential. Then let's define:

$P_w$: 3dgs center in world coord; 

$P_c$: 3dgs center in camera coord; 

$T_{lw}$: transofrmation from world to lidar coord; 

$T_{cl}$: transofrmation from lidar to camera coord (which is the extrinsic). 

$T_{c{\prime}c}$: transofrmation from wrong camera coord to correct camera coord (because we think the extrinsic has error, we can use this transformation to adjust the relative transformation from camera to lidar to correct the projection) 

$T_{w{\prime}w}$: transofrmation from wrong world coord to correct wrold coord (if you fix your camera coord(thinking it's correct), then you can adjust the 3dgs points position to correct the projection) 

$$
\begin{aligned}
P_{c{\prime}} &= T_{c{\prime}c} T_{cl} T_{lw} P_w \\
P_{c} &= T_{cl} T_{lw{\prime}} T_{w{\prime}w} P_{w}
\end{aligned}
$$

Both of these refinements ensure that the 3dgs points are correctly projected into the camera coordinates. The first one, $T_{c{\prime}c}$, is used by [MonoGS](https://github.com/muskie82/MonoGS) (if you're interested in optimizing velocity as well, you may also refer to the paper [GS-EVT](https://arxiv.org/pdf/2409.19228v1)). The second one, $T_{w{\prime}w}$, is used by [Gaussian-SLAM](https://github.com/VladimirYugay/Gaussian-SLAM). However, for the extrinsic calibration task, only the first transformation, $T_{c{\prime}c}$, should be chosen. Even if you finally get the optimized $T_{w{\prime}w}$, it cannot be used to derive the correct extrinsic $T_{cl}$. In contrast, optimizing $T_{c{\prime}c}$ allows you to directly multiply it by $T_{cl}$ to obtain the correct extrinsic: $T_{c{\prime}l} = T_{c{\prime}c} T_{cl}$.

The question is that if you want to choose the first method $T_{c{\prime}c}$, you need explicitly derive the Jacobian of 3dgs loss $L$ w.r.t. $T_{cw}$ (see [MonoGS](https://github.com/muskie82/MonoGS)), because the this part is done in CUDA, you need explicitly calculate the gradient to make maintain the back-propogation chain flow (cuda don't have autograd). If you use second method $T_{w{\prime}w}$, then you won't need do this (see [Gaussian-SLAM](https://github.com/VladimirYugay/Gaussian-SLAM)), because:

$$
\frac{\partial L}{\partial T_{w{\prime}w}} = \frac{\partial L}{\partial P_c} \frac{\partial P_c}{\partial (T_{w{\prime}w}P_w)} \frac{\partial (T_{w{\prime}w}P_w)}{\partial T_{w{\prime}w}}
$$

the $\frac{\partial L}{\partial P_c} \frac{\partial P_c}{\partial (T_{w{\prime}w}P_w)}$ is calculated by original 3dgs (they have explicitly calculated the gradient in cuda), and the $\frac{\partial (T_{w{\prime}w}P_w)}{\partial T_{w{\prime}w}} = P_w$, you can use pytorch autograd to do it.
